### PR TITLE
Add compiler command `crystal clear_cache`

### LINKED ISSUE #4730: clearing compile-time cache via crystal
```cr
# crystal --version
Crystal 0.23.1 (2017-07-16) LLVM 4.0.1
# uname -a
Darwin gilead-qc124.netel.rpi.edu 14.5.0 Darwin Kernel Version 14.5.0: Tue Apr 11 16:12:42 PDT 2017; root:xnu-2782.50.9.2.3~1/RELEASE_X86_64 x86_64
```

I was compiling a program with `crystal build -release myprog.cr`, when I realized that there was an important change which I had forgotten to make.  So I did a `ctrl-c` to interrupt the compile, and made the necessary change to my source code.

When I then redid the `crystal build -release myprog.cr`, crystal hit an error and generated a long error output.  I did not save the output and do not remember what it said, but in glancing at it I was confident that the error was not caused by the changes I had just made.  So I removed crystal's compile-time cache for that program (under `~/.cache/crystal/full_path_to_myprog.cr`), redid the build command, and this time it compiled without any errors.

But then it occurred to me, what if I had not known about that cache directory?  The error message didn't say anything about the compile-time cache or where it was.  Is there some option to `crystal` or `crystal build` or `crystal tool` which would clear that directory for the user?  What if someone wanted to have a `make clean` target, and as part of that they wanted to remove that compile-time cache directory?

### PR DESCRIPTION
**Pull request by baseballlover723**
Seemed simple enough to implement so I took a crack at it. Working perhaps isn't the best, but I think is moderately clear enough. Didn't add any tests because I don't know enough about the compiler to know how to easily generate cache files.

Fixes #4730

### CODE REVIEW

File: src/compiler/crystal/command/clear_cache.cr [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+        Options:
+        BANNER
+
+      opts.on("-h", "--help", "Show this message") do
+        puts opts
+        exit
+      end
+    end
+    puts "clearing compiler cache at \"#{CacheDir.instance.dir}\""
```
Comment (straight-shoota) [inline src/compiler/crystal/command/clear_cache.cr:19]: I think this command shouldn't need any output. Maybe make this conditional on the `--verbose` flag?

**Comment by straight-shoota** [comment]
For testing, you could just write a file to `CacheDir.instance.dir`, run the command and verify that the file is gone.
However, we don't really have much testing of compiler commands at all, so it should no be a requirement to add it for this one (it would be nice, though).

File: src/compiler/crystal/command/clear_cache.cr (hunk 1 above)
Comment (baseballlover723) [reply src/compiler/crystal/command/clear_cache.cr]: added the verbose flag.

File: src/compiler/crystal/command/clear_cache.cr [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        puts opts
+        exit
+      end
+
+      opts.on("-v", "--verbose", "Display detailed information") do
+        verbose = true
+      end
+
+    end
+    puts "clearing compiler cache at \"#{CacheDir.instance.dir}\"" if verbose
```
Comment (Sija) [inline src/compiler/crystal/command/clear_cache.cr:25]: Minor style suggestion:

```suggestion
    puts "Clearing compiler cache at #{CacheDir.instance.dir.inspect}" if verbose
```

**Comment by baseballlover723** [comment]
> For testing, you could just write a file to `CacheDir.instance.dir`, run the command and verify that the file is gone. However, we don't really have much testing of compiler commands at all, so it should no be a requirement to add it for this one (it would be nice, though).

Added a spec for it. Not 100% sure if we actually want this spec or not, since it actually deletes the actual compiler cache, and I could imagine that there might be an issue with doing that during the tests. I'm thinking that it would be problematic if the compiler tests were ever run in parallel, since it could easily interfere with another test.

I briefly took a look at making a tempdir for the test and trying to use that instead of the actual compiler cache. Couldn't reassign the CacheDir singleton, and I'm not familiar enough with crystal to really know the best way to mock out a singleton. So if you have any guidance on that then I'd be happy to modify the spec so that it uses a tempdir instead of the actual compiler cache. Otherwise, I think this is alright for now (especially since the spec isn't really a requirement for the PR)

**Comment by Sija** [comment]
> I briefly took a look at making a tempdir for the test and trying to use that instead of the actual compiler cache. Couldn't reassign the CacheDir singleton, and I'm not familiar enough with crystal to really know the best way to mock out a singleton. So if you have any guidance on that then I'd be happy to modify the spec so that it uses a tempdir instead of the actual compiler cache. Otherwise, I think this is alright for now (especially since the spec isn't really a requirement for the PR)

Closest to mocking that stdlib provides, is monkey-patching the `CacheDir` class, sth like:

```cr
class CacheDir
  class_setter instance

  def initialize(@dir)
  end
end
```

This would allow you to swap the `CacheDir` instance in specs, preferably within `around_each` hook.

**Review by straight-shoota — APPROVED** [review]
Nice. Doing that much work for the spec really wasn't expected of you.
It could potentially pose a problem if we run specs concurrently. For that to work correctly, we'd need to run the command in complete isolation. But we don't do that right now, and there are probably other issues with that as well in the compiler specs.
So I'm happy to take it in. We'll see if there happen to be any issues with it.

**Review by HertzDevil — APPROVED** [review]

**Comment by baseballlover723** [comment]
> @baseballlover723 Please avoid force pushes to a PR, as requested in the the [contributing guide](https://github.com/crystal-lang/crystal/blob/master/CONTRIBUTING.md#making-good-pull-requests). Especially, after it has been approved and milestoned.

My apologies, I didn't realize that that document existed. I'll refrain from force pushing in the future.

**Pull request merged**